tests/lib/versa_test.py

```diff
@@ -0,0 +1,476 @@
+  comment:: "Add icmp "
+  protocol:: icmp
+  action:: accept
+}
+"""
+
+ICMP_TYPE_TERM_1 = """
+term test-icmp {
+  comment:: "Add icmp type not supported"
+  protocol:: icmp
+  icmp-type:: echo-request echo-reply
+  action:: accept
+}
+"""
+
+ICMP_TYPE_TERM_2 = """
+term test-icmp {
+  comment:: "Add icmpv6 not supported"
+  protocol:: icmpv6
+  action:: accept
+}
+"""
+
+SUPPORTED_TOKENS = frozenset({
+    'action',
+    'comment',
+    'counter',
+    'destination_address',
+    'destination_address_exclude',
+    'destination_port',
+    'destination_zone',
+    'dscp_except',
+    'dscp_match',
+    'dscp_set',
+    'source_zone',
+    'expiration',
+    'icmp_type',
+    'stateless_reply',
+    'logging',
+    'name',
+    'option',
+    'owner',
+    'platform',
+    'platform_exclude',
+    'protocol',
+    'source_address',
+    'source_address_exclude',
+    'source_port',
+    'timeout',
+    'translated',
+    'verbatim',
+    'vpn'
+})
+
+SUPPORTED_SUB_TOKENS = {
+    'action': {'accept', 'deny', 'reject', 'count', 'log', 'dscp'},
+    'icmp_type': {
+        'alternate-address',
+        'certification-path-advertisement',
+        'certification-path-solicitation',
+        'conversion-error',
+        'destination-unreachable',
+        'echo-reply',
+        'echo-request',
+        'mobile-redirect',
+        'home-agent-address-discovery-reply',
+        'home-agent-address-discovery-request',
+        'icmp-node-information-query',
+        'icmp-node-information-response',
+        'information-request',
+        'inverse-neighbor-discovery-advertisement',
+        'inverse-neighbor-discovery-solicitation',
+        'mask-reply',
+        'mask-request',
+        'information-reply',
+        'mobile-prefix-advertisement',
+        'mobile-prefix-solicitation',
+        'multicast-listener-done',
+        'multicast-listener-query',
+        'multicast-listener-report',
+        'multicast-router-advertisement',
+        'multicast-router-solicitation',
+        'multicast-router-termination',
+        'neighbor-advertisement',
+        'neighbor-solicit',
+        'packet-too-big',
+        'parameter-problem',
+        'redirect',
+        'redirect-message',
+        'router-advertisement',
+        'router-renumbering',
+        'router-solicit',
+        'router-solicitation',
+        'source-quench',
+        'time-exceeded',
+        'timestamp-reply',
+        'timestamp-request',
+        'unreachable',
+        'version-2-multicast-listener-report',
+    },
+}
+
+# Print a info message when a term is set to expire in that many weeks.
+# This is normally passed from command line.
+EXP_INFO = 2
+
+_IPSET = [nacaddr.IP('10.0.0.0/8'),
+          nacaddr.IP('2001:4860:8000::/33')]
+_IPSET2 = [nacaddr.IP('10.23.0.0/22'), nacaddr.IP('10.23.0.6/23', strict=False)]
+_IPSET3 = [nacaddr.IP('10.23.0.0/23')]
+_IPSET4 = [nacaddr.IP('10.0.0.0/20')]
+_IPSET5 = [nacaddr.IP('10.0.0.0/24')]
+
+
+class VersaTest(absltest.TestCase):
```

Comment:
We probably will need to add some more unit test coverage. The generator has a ton of different options and it will help to have more cases covered - https://github.com/google/capirca/blob/master/doc/generator_patterns.md#test-coverage

Reply:
Absolutely